In [30]:
#PIP INSTALLATION
print('---1---')
%pip install pandas
print('---2---')
%pip install seaborn
print('---3---')
%pip install scikit-learn
print('---4---')
%pip install catboost
print('---5---')
%pip install ipywidgets
print('---6---')
%pip install pandas openpyxl


---1---
Note: you may need to restart the kernel to use updated packages.
---2---
Note: you may need to restart the kernel to use updated packages.
---3---
---4---

Note: you may need to restart the kernel to use updated packages.
---5---
Note: you may need to restart the kernel to use updated packages.
---6---
Note: you may need to restart the kernel to use updated packages.


In [31]:
#FROM LIBRARIES
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from catboost import CatBoostClassifier, Pool
from IPython.display import FileLink
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score

In [32]:
#IMPORT AS
import numpy as np
import pandas as pd
import seaborn as sns

import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('ggplot')


In [33]:
file_path = 'Dataset/diabetic_data.csv'
df = pd.read_csv(file_path,keep_default_na=False)

In [34]:
df_clean = df.copy()

In [35]:
df_clean.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

In [36]:
df_clean.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

In [37]:
print(list(df_clean))

['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted']


Data Cleaning 

In [38]:
#REMOVING DUPLICATE DATA'S
df_clean['patient_nbr'].value_counts()

patient_nbr
88785891     40
43140906     28
1660293      23
88227540     23
23199021     23
             ..
11005362      1
98252496      1
1019673       1
13396320      1
175429310     1
Name: count, Length: 71518, dtype: int64

In [39]:
df_clean = df_clean.drop_duplicates(subset=['patient_nbr'])
#END REMOVING DUPLICATE DATA'S

There are 30,248 total of Duplicate Data's which the unique data consist only 71,518

In [40]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71518 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   encounter_id              71518 non-null  int64 
 1   patient_nbr               71518 non-null  int64 
 2   race                      71518 non-null  object
 3   gender                    71518 non-null  object
 4   age                       71518 non-null  object
 5   weight                    71518 non-null  object
 6   admission_type_id         71518 non-null  int64 
 7   discharge_disposition_id  71518 non-null  int64 
 8   admission_source_id       71518 non-null  int64 
 9   time_in_hospital          71518 non-null  int64 
 10  payer_code                71518 non-null  object
 11  medical_specialty         71518 non-null  object
 12  num_lab_procedures        71518 non-null  int64 
 13  num_procedures            71518 non-null  int64 
 14  num_medications           

Value of each drug that most patient uses

In [53]:
print(df['metformin'].value_counts())
print("-----------------------------")
print(df['repaglinide'].value_counts())
print("-----------------------------")
print(df['nateglinide'].value_counts())
print("-----------------------------")
print(df['chlorpropamide'].value_counts())

metformin
No        81778
Steady    18346
Up         1067
Down        575
Name: count, dtype: int64
-----------------------------
repaglinide
No        100227
Steady      1384
Up           110
Down          45
Name: count, dtype: int64
-----------------------------
nateglinide
No        101063
Steady       668
Up            24
Down          11
Name: count, dtype: int64
-----------------------------
chlorpropamide
No        101680
Steady        79
Up             6
Down           1
Name: count, dtype: int64


In [52]:
print(df['glimepiride'].value_counts())
print("-----------------------------")
print(df['acetohexamide'].value_counts())
print("-----------------------------")
print(df['glipizide'].value_counts())
print("-----------------------------")
print(df['glyburide'].value_counts())


glimepiride
No        96575
Steady     4670
Up          327
Down        194
Name: count, dtype: int64
-----------------------------
acetohexamide
No        101765
Steady         1
Name: count, dtype: int64
-----------------------------
glipizide
No        89080
Steady    11356
Up          770
Down        560
Name: count, dtype: int64


In [54]:
print(df['tolbutamide'].value_counts())
print("-----------------------------")
print(df['pioglitazone'].value_counts())
print("-----------------------------")
print(df['rosiglitazone'].value_counts())
print("-----------------------------")
print(df['acarbose'].value_counts())

tolbutamide
No        101743
Steady        23
Name: count, dtype: int64
-----------------------------
pioglitazone
No        94438
Steady     6976
Up          234
Down        118
Name: count, dtype: int64
-----------------------------
rosiglitazone
No        95401
Steady     6100
Up          178
Down         87
Name: count, dtype: int64
-----------------------------
acarbose
No        101458
Steady       295
Up            10
Down           3
Name: count, dtype: int64


In [55]:
print(df['miglitol'].value_counts())
print("-----------------------------")
print(df['troglitazone'].value_counts())
print("-----------------------------")
print(df['tolazamide'].value_counts())
print("-----------------------------")
print(df['examide'].value_counts())

miglitol
No        101728
Steady        31
Down           5
Up             2
Name: count, dtype: int64
-----------------------------
troglitazone
No        101763
Steady         3
Name: count, dtype: int64
-----------------------------
tolazamide
No        101727
Steady        38
Up             1
Name: count, dtype: int64
-----------------------------
examide
No    101766
Name: count, dtype: int64


In [56]:
print(df['citoglipton'].value_counts())
print("-----------------------------")
print(df['insulin'].value_counts())
print("-----------------------------")
print(df['glyburide-metformin'].value_counts())
print("-----------------------------")
print(df['glipizide-metformin'].value_counts())

citoglipton
No    101766
Name: count, dtype: int64
-----------------------------
insulin
No        47383
Steady    30849
Down      12218
Up        11316
Name: count, dtype: int64
-----------------------------
glyburide-metformin
No        101060
Steady       692
Up             8
Down           6
Name: count, dtype: int64
-----------------------------
glipizide-metformin
No        101753
Steady        13
Name: count, dtype: int64


In [57]:
print(df['glimepiride-pioglitazone'].value_counts())
print("-----------------------------")
print(df['metformin-rosiglitazone'].value_counts())
print("-----------------------------")
print(df['metformin-pioglitazone'].value_counts())

glimepiride-pioglitazone
No        101765
Steady         1
Name: count, dtype: int64
-----------------------------
metformin-rosiglitazone
No        101764
Steady         2
Name: count, dtype: int64
-----------------------------
metformin-pioglitazone
No        101765
Steady         1
Name: count, dtype: int64


Removing features that is not useful for our target readmittied

encounter_id  => Irrelevant 

Patient_nbr => Irrelevant since this is only the ID of the patient

Weight => Consist of 97% missing value

Payer_code => Irrelevant since this only a payment method

... to be Continue

In [11]:
features_drop_list = ['encounter_id', 'patient_nbr', 'weight', 'payer_code', 'medical_specialty', 'repaglinide', 'nateglinide', 'chlorpropamide', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone','metformin-pioglitazone', 'acetohexamide', 'tolbutamide']


In [12]:
df_clean.drop(features_drop_list, axis=1, inplace=True)



In [13]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71518 entries, 0 to 101765
Data columns (total 29 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   race                      71518 non-null  object
 1   gender                    71518 non-null  object
 2   age                       71518 non-null  object
 3   admission_type_id         71518 non-null  int64 
 4   discharge_disposition_id  71518 non-null  int64 
 5   admission_source_id       71518 non-null  int64 
 6   time_in_hospital          71518 non-null  int64 
 7   num_lab_procedures        71518 non-null  int64 
 8   num_procedures            71518 non-null  int64 
 9   num_medications           71518 non-null  int64 
 10  number_outpatient         71518 non-null  int64 
 11  number_emergency          71518 non-null  int64 
 12  number_inpatient          71518 non-null  int64 
 13  diag_1                    71518 non-null  object
 14  diag_2                    

Understand the Diagnosis...

In [14]:
#start by setting all values containing E or V into 0 (as one category)
df_clean.loc[df_clean['diag_1'].str.contains('V',na=False,case=False), 'diag_1'] = 0
df_clean.loc[df_clean['diag_1'].str.contains('E',na=False,case=False), 'diag_1'] = 0
df_clean.loc[df_clean['diag_2'].str.contains('V',na=False,case=False), 'diag_2'] = 0
df_clean.loc[df_clean['diag_2'].str.contains('E',na=False,case=False), 'diag_2'] = 0
df_clean.loc[df_clean['diag_3'].str.contains('V',na=False,case=False), 'diag_3'] = 0
df_clean.loc[df_clean['diag_3'].str.contains('E',na=False,case=False), 'diag_3'] = 0

#setting all missing values into -1
df_clean['diag_1'] = df_clean['diag_1'].replace('?', -1)
df_clean['diag_2'] = df_clean['diag_2'].replace('?', -1)
df_clean['diag_3'] = df_clean['diag_3'].replace('?', -1)

#No all diag values can be converted into numeric values
df_clean['diag_1'] = df_clean['diag_1'].astype(float)
df_clean['diag_2'] = df_clean['diag_2'].astype(float)
df_clean['diag_3'] = df_clean['diag_3'].astype(float)

In [15]:
print(df_clean.head())

              race  gender      age  admission_type_id  \
0        Caucasian  Female   [0-10)                  6   
1        Caucasian  Female  [10-20)                  1   
2  AfricanAmerican  Female  [20-30)                  1   
3        Caucasian    Male  [30-40)                  1   
4        Caucasian    Male  [40-50)                  1   

   discharge_disposition_id  admission_source_id  time_in_hospital  \
0                        25                    1                 1   
1                         1                    7                 3   
2                         1                    7                 2   
3                         1                    7                 2   
4                         1                    7                 1   

   num_lab_procedures  num_procedures  num_medications  ...  metformin  \
0                  41               0                1  ...         No   
1                  59               0               18  ...         No   
2       

In [17]:
df_clean['race'] = df_clean['race'].replace('?', 'Other')

In [18]:
df_clean.race.value_counts()

race
Caucasian          53491
AfricanAmerican    12887
Other               3126
Hispanic            1517
Asian                497
Name: count, dtype: int64

In [19]:
print(df_clean['race'].isna())

0         False
1         False
2         False
3         False
4         False
          ...  
101754    False
101755    False
101756    False
101758    False
101765    False
Name: race, Length: 71518, dtype: bool


In [20]:
df_clean.gender.value_counts()

gender
Female             38025
Male               33490
Unknown/Invalid        3
Name: count, dtype: int64

In [21]:
df_clean['gender'] = df_clean['gender'].replace('Unknown/Invalid', 'Female')
df_clean.gender.value_counts()

gender
Female    38028
Male      33490
Name: count, dtype: int64

In [22]:
df_clean.age.value_counts()


age
[70-80)     18210
[60-70)     15960
[50-60)     12466
[80-90)     11589
[40-50)      6878
[30-40)      2699
[90-100)     1900
[20-30)      1127
[10-20)       535
[0-10)        154
Name: count, dtype: int64

In [23]:
for i in range(0, 10):
    age_range = f'[{10*i}-{10*(i+1)})'
    midpoint = i*10 + 5
    df['age'] = df['age'].replace(age_range, midpoint)

C:\Users\SnowPlay\AppData\Local\Temp\ipykernel_22472\3365448648.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['age'] = df['age'].replace(age_range, midpoint)


In [24]:
df_clean['age'].value_counts()

age
[70-80)     18210
[60-70)     15960
[50-60)     12466
[80-90)     11589
[40-50)      6878
[30-40)      2699
[90-100)     1900
[20-30)      1127
[10-20)       535
[0-10)        154
Name: count, dtype: int64

In [25]:
df_clean.max_glu_serum.value_counts()


max_glu_serum
None    68062
Norm     1731
>200      969
>300      756
Name: count, dtype: int64

In [26]:
label_encoder = LabelEncoder()
df_clean['readmitted'] = label_encoder.fit_transform(df_clean['readmitted'])

In [27]:
X = df_clean.drop('readmitted', axis=1)  # Features
y = df_clean['readmitted']               # Target variable


In [28]:
cat_features = [index for index, dtype in enumerate(X.dtypes) if dtype == 'object']


In [29]:
print(X.iloc[:, cat_features].isnull().sum())


race             0
gender           0
age              0
max_glu_serum    0
A1Cresult        0
metformin        0
glimepiride      0
glipizide        0
glyburide        0
pioglitazone     0
rosiglitazone    0
insulin          0
change           0
diabetesMed      0
dtype: int64


In [30]:
# Save as Excel
df_clean.to_excel('clean_dataframe.xlsx', index=False, engine='openpyxl')

In [31]:
FileLink('clean_dataframe.xlsx')

D:\Optimizing Catboost\Thesis Application\clean_dataframe.xlsx

Training Model

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [33]:
train_pool = Pool(data=X_train, label=y_train, cat_features=cat_features)
test_pool = Pool(data=X_test, label=y_test, cat_features=cat_features)


In [34]:
# Step 5: Train the CatBoost model
model = CatBoostClassifier(
    loss_function = 'MultiClass',
    iterations=1000,          # Number of boosting rounds
    learning_rate=0.03,       # Step size shrinkage
    depth=6,                  # Depth of trees
    verbose=100               # Display training logs
)


In [35]:
model.fit(train_pool)

0:	learn: 1.0856792	total: 253ms	remaining: 4m 12s
100:	learn: 0.8493245	total: 10.6s	remaining: 1m 34s
200:	learn: 0.8374526	total: 22s	remaining: 1m 27s
300:	learn: 0.8315360	total: 33.4s	remaining: 1m 17s
400:	learn: 0.8261400	total: 44.9s	remaining: 1m 7s
500:	learn: 0.8201530	total: 55.7s	remaining: 55.5s
600:	learn: 0.8149283	total: 1m 6s	remaining: 44.2s
700:	learn: 0.8103947	total: 1m 17s	remaining: 33.1s
800:	learn: 0.8067159	total: 1m 28s	remaining: 22s
900:	learn: 0.8035194	total: 1m 39s	remaining: 11s
999:	learn: 0.8004393	total: 1m 50s	remaining: 0us


In [36]:
# Step 9: Make predictions
y_pred = model.predict(X_test)

In [37]:
# Step 10: Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy: {accuracy:.2f}")

Model Accuracy: 0.63


In [38]:
# Additional step: Visualize feature importance
feature_importances = model.get_feature_importance()
feature_names = X.columns
importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': feature_importances})
importance_df = importance_df.sort_values(by='Importance', ascending=False)

print(importance_df)

                     Feature  Importance
4   discharge_disposition_id   20.379270
13                    diag_1    8.012305
12          number_inpatient    7.195950
14                    diag_2    6.232519
2                        age    5.428857
15                    diag_3    5.289500
16          number_diagnoses    4.756848
7         num_lab_procedures    4.712187
5        admission_source_id    4.605473
3          admission_type_id    3.986480
6           time_in_hospital    3.938483
9            num_medications    3.908597
8             num_procedures    3.078356
10         number_outpatient    2.546531
11          number_emergency    2.203625
25                   insulin    1.922001
0                       race    1.779003
27               diabetesMed    1.550935
18                 A1Cresult    1.468835
19                 metformin    1.255925
1                     gender    1.084707
21                 glipizide    0.917814
23              pioglitazone    0.808475
17             m

In [39]:
model.save_model("catboost_model.cbm", format="cbm")